proof_ipa.go

```diff
@@ -79,36 +82,76 @@ func GetCommitmentsForMultiproof(root VerkleNode, keys [][]byte, resolver NodeRe
+// a proof.
+func getProofElementsFromTree(preroot, postroot VerkleNode, keys [][]byte, resolver NodeResolverFn) (*ProofElements, []byte, [][]byte, [][]byte, []*Point, [][]Fr, []*Fr, []byte, error) {
 	// go-ipa won't accept no key as an input, catch this corner case
 	// and return an empty result.
 	if len(keys) == 0 {
-		return nil, nil, nil, nil, errors.New("no key provided for proof")
+		return nil, nil, nil, nil, nil, nil, nil, nil, errors.New("no key provided for proof")
 	}
 
-	tr := common.NewTranscript("vt")
-	root.Commit()
-
-	pe, es, poas, err := GetCommitmentsForMultiproof(root, keys, resolver)
+	pe, es, poas, err := GetCommitmentsForMultiproof(preroot, keys, resolver)
 	if err != nil {
-		return nil, nil, nil, nil, err
+		return nil, nil, nil, nil, nil, nil, nil, nil, fmt.Errorf("error getting pre-state proof data: %w", err)
+	}
+
+	// List of points and vectors and indices used to generate the IPA proof
+	// If only a pre-state root is available, it's a simple copy, but if both
+	// pre- and post-state root are present, it's merging the two lists.
+	var (
+		proof_cis []*Point
+		proof_fs  [][]Fr
+		proof_zs  []uint8
+		proof_ys  []*Fr
+		postvals  = make([][]byte, len(keys))
+	)
+	for i := range pe.Cis {
+		proof_cis = append(proof_cis, pe.Cis[i])
+		proof_fs = append(proof_fs, pe.Fis[i])
+		proof_ys = append(proof_ys, pe.Yis[i])
+		proof_zs = append(proof_zs, pe.Zis[i])
+	}
+
+	// if a post-state tree is present, merge its proof elements with
+	// those of the pre-state tree, so that they can be proved together.
+	if postroot != nil {
+		pe_post, _, _, err := GetCommitmentsForMultiproof(postroot, keys, resolver)
+		if err != nil {
+			return nil, nil, nil, nil, nil, nil, nil, nil, fmt.Errorf("error getting post-state proof data: %w", err)
+		}
+
+		for i := range pe_post.Cis {
+			proof_cis = append(proof_cis, pe_post.Cis[i])
+			proof_fs = append(proof_fs, pe_post.Fis[i])
+			proof_ys = append(proof_ys, pe_post.Yis[i])
+			proof_zs = append(proof_zs, pe_post.Zis[i])
+		}
+
+		// Set the post values, if they are untouched, leave them `nil`
+		for i, v := range pe.Vals {
+			if !bytes.Equal(v, pe_post.Vals[i]) {
+				postvals[i] = v
+			}
+		}
 	}
 
-	// NOTE this is leftover code from the time the proof was
-	// made against the POST state. Since proofs are expected
-	// to prove PRE and POST state in the future, I'm leaving
-	// this for reference - eventhough it's unlikely that the
-	// final version will look like this, but you never know.
-	// var vals [][]byte
-	// for _, k := range keys {
-	// 	// TODO at the moment, do not include the post-data
-	// 	// val, _ := root.Get(k, nil)
-	// 	// vals = append(vals, val)
-	// 	vals = append(vals, keyvals[string(k)])
-	// }
+	// 1..=3: proof elements of the pre-state trie for serialization,
+	// 4: values to be inserted in the post-state trie for serialization
+	// 5..=7: aggregated pre+post elements for proving
```

Comment:
Nit: `5..=8`

Instant karma for non-idiomatic syntax :laughing:

Reply:
ok, I'll use the zig syntax